Lib/argparse.py

```diff
@@ -1859,6 +1845,11 @@ def parse_known_args(self, args=None, namespace=None):
         else:
             namespace, args = self._parse_known_args(args, namespace)
 
+        # add any parser defaults that aren't present
+        for dest in self._defaults:
+            if not hasattr(namespace, dest):
+                setattr(namespace, dest, self._defaults[dest])
```

Comment:
The default value we are referring to in an `Action` doesn't necessarily comes from an argument like `--foo`. It may come from [`set_defaults`](https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.set_defaults):

```py
            command_parser.set_defaults(
                func=metadata,
                command=command_name,
                _cmd=metadata,
```

Thus, when calling `self._parse_known_args`, `self._defaults` will be in `namespace`.

![image](https://user-images.githubusercontent.com/4003950/145921199-68fce273-17a4-4506-b7fb-8881bbc4950c.png)

If these 3 lines are moved down here, `self._defaults` won't be in `namespace` when calling `self._parse_known_args`, resulting in failure

```
  File "C:\Users\name\AppData\Local\Programs\Python\Python310\lib\argparse.py", line 2003, in consume_optional
    take_action(action, args, option_string)
  File "C:\Users\name\AppData\Local\Programs\Python\Python310\lib\argparse.py", line 1931, in take_action
    action(self, namespace, argument_values, option_string)
  File "d:\cli\azure-cli\src\azure-cli-core\azure\cli\core\commands\arm.py", line 357, in __call__
    profile = Profile(cli_ctx=namespace._cmd.cli_ctx)  # pylint: disable=protected-access
AttributeError: 'Namespace' object has no attribute '_cmd'
```

I am open to whether this is the correct behavior, but it is definitely a BREAKING CHANGE given the behavior has been there for years. If this change is what Python committee want, it should be released in at least Python 3.12, not Python 3.10 or 3.11 which are already stabilized.

Reply:
This case makes sense to me. I admit that any change may be a breaking change to some people. This needs more discussion, it perhaps should happen on BPO I guess?

Reply:
Yes, these are great concerns to get feedback from BPO-45235. Some backward compatibility concerns are already in that thread, and as maintainer of another project that was affected by the reverted change in Python 3.9.8, I'd love to see those get hashed out in the main thread instead of the line notes of this patch.

Reply:
Could you in the meantime submit patches for the unit test we’re missing?

There are at least two - for the regression that happened and the one that was identified here.

Reply:
they are included in the one added case, do you suggest to separate into two cases?

Reply:
I'm going to step away from this one.   The last attempt at a fix was catastrophic and should teach us that almost every nuance of the implementation is being relied upon.   I think you're correct in saying "any change may be a breaking change to some people. "  ISTM that there is no possible benefit that would be worth breaking some long deployed, stable, working command-line tools

Reply:
In my second attempt #30219 the side effect is removed, it should be cleaner.

The issue itself is a valid one, but the first fix isn't that careful.